IDP integration (#128)
percona/everest-doc: docs/secure/Idp_integration.md

[Identity Provider (IdP)](https://www.okta.com/identity-101/why-your-company-needs-an-identity-provider/) integration connects applications and services with an external identity provider for your organization. This enables centralized authentication and authorization management, improving security and simplifying user access. By leveraging IdP integration, you can ensure that users are securely authenticated and authorized to access various applications and services across your organization.

Percona Everest uses [OpenID Connect](https://auth0.com/docs/authenticate/protocols/openid-connect-protocol) (OIDC) Protocol to integrate with external Identity Providers (IdP).

## Configure OIDC on the providers's side

- **PKCE**: When setting up the Provider side, it's important to configure an application specifically for [PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow#authorization-code-flow-with-proof-key-for-code-exchange-pkce-) (Proof Key for Code Exchange) authorization.

    !!! info "Important"
        The login redirect URL should point to `<EVEREST_URL>/login-callback`, and the logout redirect URL to `<EVEREST_URL>`.

> - **https based**: IdP providers often require a secure connection (https). Therefore, the provider might require your `<EVEREST_URL>` to be based on https.

Reviewer: Also, I think this should be a info note instead of a bullet point, see the above comment.

> - **https based**: IdP providers often require a secure connection (https). Therefore, the provider might require your `<EVEREST_URL>` to be based on https.

Author: fixed

        1. Sign in to your Okta organization as a user with administrative privileges.
        
        2. In the **Admin** Console, go to **Applications → Applications** and click **Create App Integration**.

        3. On the **Create a new app integration page**, set the following:
        
            - Sign-in method - **OIDC - OpenID Connect**
            - **Application type** - **Signle-Page Application**, and click **Next**.

            ![!image](../images/OKTA_admin_console.png)

        4. Set the following fields:

            a. App integration name - any value

            b. Sign-in redirect URIs - `<EVEREST_URL>/login-callback`

            c. Sign-out redirect URIs - `<EVEREST_URL>`

            d. Click **Save**.

            e. Copy the `clientID` of the created app.

            f. Navigate to **Security → API → Authorization Servers** and copy the `issuerURL` you’d like to use for the Everest authorization. 

            !!! note "Note"
                Okta allows the use of HTTP for development purposes and in cases where the Admin explicitly permits it.

            ![!image](../images/sso_aap_integration.png)

## Configure OIDC on Percona Everest side

> You can configure OIDC on the Percona side using the Percona Everest CLI or the wizard.

Reviewer 2: A bit confusing here. In fact _both options_ use the Percona Everest CLI to configure OIDC, it's even the same command `everestctl settings oidc configure`. The point is there are two ways how to provide the configuration itself - by passing the flags or by following the wizard.

Author: Got it.

Configuring OIDC using the flag:

    everestctl settings oidc configure --issuer-url=http://url.com --client-id=<your-app-client-id

??? example "Output"

    ```{.text .no-copy}
    2024-06-18T11:06:18Z    info    oidc/configure.go:110   OIDC provider configured, restarting Everest..
    2024-06-18T11:06:33Z    info    oidc/configure.go:117   OIDC has been configured successfully
    ```

Configuring OIDC via wizard:

??? example "Output"

    ```{.text .no-copy}
    $ everestctl settings oidc configure
    ? Enter issuer URL <your-provider-url>
    ? Enter client ID <your-app-client-id>
    2024-06-18T11:05:15Z    info    oidc/configure.go:110   OIDC provider   configured, restarting Everest..    
    2024-06-18T11:05:30Z    info    oidc/configure.go:117   OIDC has been configured successfully
    ```

> ??? info "What's happening under the hood?"

Reviewer: :clap: I really like this

    This command stores the updated configuration in k8s and restarts the Everest deployment.

    1. **Store Configuration**

>         The OIDC settings are stored in the `everest-settings` ConfigMap, along with other settings, in the following format:

Reviewer: ```suggestion
        The OIDC settings are stored in the `everest-settings` ConfigMap in the `everest-system` namespace, along with other settings, in the following format:
```

        ```sh
        apiVersion: v1
        kind: ConfigMap
        metadata:

>             name: everest-settings

Reviewer: ```suggestion
            name: everest-settings
            namespace: everest-system
```

    2. **Restart Percona Everest**

        OIDC configuration must be in place for Percona Everest to start the API Server, allowing the server to utilize the OIDC provider's validation. This means that after setting up the OIDC configuration, the Percona Everest API Server needs to be restarted.

        The CLI command typically takes approximately 15 seconds to execute. It waits for the Everest Deployment to be up again before exiting successfully.

    
        !!! info "Important"

            - The restart only impacts the Everest UI and API. Database clusters are not affected.

            - The restart results in the loss of the port-forwarding connection. If you had port-forwarding enabled on your machine to access Percona Everest UI and API, you will need to set it up again.

## API to get the OIDC settings

The OIDC configuration settings can be retrieved along with other Everest settings using the following API:

```sh
GET /v1/settings
```

??? example "Example"

        {        
           
            "oidcConfig": {
            
            "clientId": "id1234",
            
            "issuerURL": "https://url.com"
        
            }
    
        }

> For more information, see the [Open api spec](docs/spec/openapi.yml).

Reviewer 2: I don't think we should add this section to the docs, @recharte wdyt? The end users don't need the API, because the UI does all the job for them. Maybe it would be better to add a section of what will happen next after the oidc is configured successfully - that the users should go to the Everest login page and they should see that "login with SSO" button.

Reviewer: I agree, API has it's own separate docs I don't think it makes sense to have it here.

Author: Fixed.

        
        2. In the **Admin** Console, go to **Applications → Applications** and click **Create App Integration**.

        3. On the **Create a new app integration page**, set the following:
        
            - Sign-in method - **OIDC - OpenID Connect**
            - **Application type** - **Signle-Page Application**, and click **Next**.

            ![!image](../images/OKTA_admin_console.png)

        4. Set the following fields:

            a. App integration name - any value

            b. Sign-in redirect URIs - `<EVEREST_URL>/login-callback`

            c. Sign-out redirect URIs - `<EVEREST_URL>`

            d. Click **Save**.

            e. Copy the `clientID` of the created app.

            f. Navigate to **Security → API → Authorization Servers** and copy the `issuerURL` you’d like to use for the Everest authorization. 

            !!! note "Note"
                Okta allows the use of HTTP for development purposes and in cases where the Admin explicitly permits it.

            ![!image](../images/sso_aap_integration.png)

## Configure OIDC on Percona Everest side

You can configure OIDC via Percona Everest CLI:

>     everestctl settings oidc configure --issuer-url=http://url.com --client-id=id2342

Reviewer 2: Just for consideration.  
There are two options how we can run the config - using flags (as described here) or using wizard. 

The flags option's output looks like:
```
$ everestctl settings oidc configure --issuer-url=<your-provider-url> --client-id=<your-app-client-id>
2024-06-18T11:06:18Z    info    oidc/configure.go:110   OIDC provider configured, restarting Everest..
2024-06-18T11:06:33Z    info    oidc/configure.go:117   OIDC has been configured successfully
```

The wizard option looks like:
```
$ everestctl settings oidc configure                                     
? Enter issuer URL <your-provider-url>
? Enter client ID <your-app-client-id>
2024-06-18T11:05:15Z    info    oidc/configure.go:110   OIDC provider configured, restarting Everest..
2024-06-18T11:05:30Z    info    oidc/configure.go:117   OIDC has been configured successfully
```

Author: Fixed

Percona Everest uses [OpenID Connect](https://auth0.com/docs/authenticate/protocols/openid-connect-protocol){:target="_blank"} (OIDC) Protocol to integrate with external Identity Providers (IdP).

!!! note "Note"
    To integrate IdP with Percona Everest, first, [install Percona Everest](../install/installEverest.md) and then [configure OIDC](#configure-oidc-on-the-providerss-side) on the IdP's side as well as the [Percona Everest side](#configure-oidc-on-percona-everest-side).

## Configure OIDC on the providers's side

- **Proof Key for Code Exchange (PKCE)**: When setting up the Provider side, it's important to configure an application specifically for [PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow#authorization-code-flow-with-proof-key-for-code-exchange-pkce-){:target="_blank"} authorization.

- **Redirect URIs**: 

    * Sign-in redirect URIs should point to `<EVEREST_URL>/login-callback`.
    * Sign-out redirect URIs should point to `<EVEREST_URL>`.
 
        !!! info "Important" 

>             IdP providers often require a secure connection (Https). Therefore, the         provider might require your `<EVEREST_URL>` to be based on Https. If your provider   requires this, consider adding a reverse proxy in front of Percona Everest to provide such functionality.

Reviewer: HTTPS is an acronym, let's use it with CAPS.
```suggestion
            IdP providers often require a secure connection (HTTPS). Therefore, the provider might require your `<EVEREST_URL>` to be based on HTTPS. If your provider requires this, consider adding a reverse proxy in front of Percona Everest to provide such functionality.
```

??? example "OKTA"

    1. Sign in to your Okta organization as a user with administrative privileges.
        
    2. In the **Admin** Console, go to **Applications → Applications** and click **Create App Integration**.

    3. On the **Create a new app integration page**, set the following:
        
        - Sign-in method - **OIDC - OpenID Connect**
        - **Application type** - **Signle-Page Application**, and click **Next**.

        ![!image](../images/OKTA_admin_console.png)

    4. Set the following fields:

        a. App integration name - any value

        b. Sign-in redirect URIs - `<EVEREST_URL>/login-callback`

        c. Sign-out redirect URIs - `<EVEREST_URL>`

        d. Click **Save**.

        e. Copy the `clientID` of the created app.

        f. Navigate to **Security → API → Authorization Servers** and copy the `issuerURL` you’d like to use for the Everest authorization. 

        !!! note "Note"
            Okta allows the use of HTTP for development purposes and in cases where the Admin explicitly permits it.

        ![!image](../images/sso_aap_integration.png)

> ## Configure OIDC on Percona Everest side

Reviewer: ```suggestion
## Configure OIDC on the Percona Everest's side
```

> You can configure OIDC on the Percona side using the flag or the wizard.

Reviewer: Maybe we should have a quick mention about the Issuer URL and Client ID. We should mention that these were the fields we asked the user to copy in the previous section.

> You can configure OIDC on the Percona side using the flag or the wizard.

Reviewer: ```suggestion
You can configure OIDC on the Percona Everest's side using the CLI in headless mode or with the wizard.
```
